config/configload/load.go

```diff
@@ -48,218 +51,255 @@ func LoadBytes(src []byte, filename string) (*config.CouperFile, error) {
+	if err != nil {
+		return nil, err
 	}
 
-	// reading per server block and merge backend settings which results in a final server configuration.
 	for _, serverBlock := range content.Blocks.OfType(server) {
-		srv := &config.Server{}
-		if diags = gohcl.DecodeBody(serverBlock.Body, file.Context, srv); diags.HasErrors() {
+		serverConfig := &config.Server{}
+		if diags = gohcl.DecodeBody(serverBlock.Body, couperConfig.Context, serverConfig); diags.HasErrors() {
 			return nil, diags
 		}
 
-		if len(serverBlock.Labels) > 0 {
-			srv.Name = serverBlock.Labels[0]
-		}
-
-		file.Server = append(file.Server, srv)
-
-		serverBodies, err := mergeBackendBodies(backends, srv)
+		serverBody, err := mergeBodys(backends, serverConfig)
 		if err != nil {
 			return nil, err
 		}
-		srv.Remain = MergeBodies(serverBodies)
 
-		// api block(s)
-		for _, apiBlock := range srv.APIs {
-			if apiBlock == nil {
-				continue
-			}
+		serverConfig.Remain = serverBody
 
-			bodies, err := mergeBackendBodies(backends, apiBlock)
+		for i, apiConfig := range serverConfig.APIs {
+			apiBody, err := mergeBodys(backends, apiConfig)
 			if err != nil {
 				return nil, err
 			}
 
-			bodies = appendUniqueBodies(serverBodies, bodies...)
+			apiBody = MergeBodies([]hcl.Body{serverBody, apiBody})
 
-			// empty bodies would be removed with a hcl.Merge.. later on.
-			if err = refineEndpoints(backends, bodies, apiBlock.Endpoints); err != nil {
+			serverConfig.APIs[i].Remain = apiBody
+
+			err = refineEndpoints(backends, apiBody, apiConfig.Endpoints)
+			if err != nil {
 				return nil, err
 			}
 		}
 
 		// standalone endpoints
-		if err := refineEndpoints(backends, serverBodies, srv.Endpoints); err != nil {
+		err = refineEndpoints(backends, serverBody, serverConfig.Endpoints)
+		if err != nil {
 			return nil, err
 		}
+
+		couperConfig.Servers = append(couperConfig.Servers, serverConfig)
 	}
 
-	if len(file.Server) == 0 {
+	if len(couperConfig.Servers) == 0 {
 		return nil, fmt.Errorf("configuration error: missing server definition")
 	}
-	return file, nil
+
+	return couperConfig, nil
 }
 
-func mergeBackendBodies(backendList Backends, inlineBackend config.Inline) ([]hcl.Body, error) {
-	reference, err := backendList.WithName(inlineBackend.Reference())
+// getReference tries to fetch a backend from `definitions`
+// block by a reference name, e.g. `backend = "name"`.
+func getReference(
+	definitionsBackends Backends, inline config.Inline,
+) (hcl.Body, error) {
+	reference, err := definitionsBackends.WithName(inline.Reference())
 	if err != nil {
-		r := inlineBackend.Body().MissingItemRange()
+		// Backend reference is given, but not defined in definitions.
+		r := inline.HCLBody().MissingItemRange()
 		err.(hcl.Diagnostics)[0].Subject = &r
+	}
+
+	if reference == nil {
+		reference = EmptyBody()
+	}
+
+	return reference, err
+}
+
+// mergeBodys guarantees to return a non-nil <hcl.Body> in non-error cases.
+func mergeBodys(
```

Comment:
This adverse rename should be reverted. Since this method still works on backends and there are two mergeBodies methods now (public/private) with an entirely different behaviour.